TaskLayer/XLSearchTask/XLSearchTask.cs

```diff
@@ -125,22 +130,61 @@ protected override MyTaskResults RunSpecific(string OutputFolder, List<DbForTask
 
             ReportProgress(new ProgressEventArgs(100, "Done with all searches!", new List<string> { taskId, "Individual Spectra Files" }));
 
-            var allPsmsList = allPsms.Select(p => p.First()).OrderByDescending(p => p.XLTotalScore).ToList();
+            List<List<CrosslinkSpectralMatch>> filteredAllPsms = new List<List<CrosslinkSpectralMatch>>();
 
-            foreach (var psm in allPsmsList)
+            foreach (var psmsPerScan in allPsms)
             {
-                psm.ResolveAllAmbiguities();
-                if (psm.BetaPeptide != null)
+                foreach (var psm in psmsPerScan)
                 {
-                    psm.BetaPeptide.ResolveAllAmbiguities();
+                    psm.ResolveAllAmbiguities();
+                    if (psm.BetaPeptide != null)
+                    {
+                        psm.BetaPeptide.ResolveAllAmbiguities();
+                    }
                 }
+                filteredAllPsms.Add(RemoveDuplicateFromPsmsPerScan(psmsPerScan));
             }
 
+            var allPsmsList = filteredAllPsms.Select(p => p.First()).OrderByDescending(p => p.XLTotalScore).ToList();
+
             PostXLSearchAnalysisTask postXLSearchAnalysisTask = new PostXLSearchAnalysisTask();
 
             return postXLSearchAnalysisTask.Run(OutputFolder, dbFilenameList, currentRawFileList, taskId, fileSettingsList, allPsmsList, CommonParameters, XlSearchParameters, proteinList, variableModifications, fixedModifications, localizeableModificationTypes, MyTaskResults);
 
         }
 
+        //Remove same peptide with from different protein.
+        public static List<CrosslinkSpectralMatch> RemoveDuplicateFromPsmsPerScan(List<CrosslinkSpectralMatch> crosslinkSpectralMatches)
+        {
+            Dictionary<string, CrosslinkSpectralMatch> keyValuePairs = new Dictionary<string, CrosslinkSpectralMatch>();
+            foreach (var csm in crosslinkSpectralMatches)
+            {
+                if (csm == null)
+                {
+                    continue;
+                }
+                string betaFulseq = "-";
+                if (csm.BetaPeptide != null)
+                {
+                    betaFulseq += csm.BetaPeptide.FullSequence;
+                }
+
+                if (keyValuePairs.ContainsKey(csm.FullSequence + betaFulseq))
```

Comment:
unit test

Reply:
Covered in unit test TestMixedMs2Ms2()''